### PR TITLE
Add wrapX option for ol.source.WMTS

### LINKED ISSUE #3386: Store width for ol.tilegrid.TileGrid
For wrapping the x-axis horizontally, a tile grid's extent is not sufficiently defined by providing its origin. We also need its width in tiles. This information is also advertised e.g. in WMTS capabilities.

Once we have that, we can use the width of the tile grid instead of relying on the width of a tile range calculated from an unrelated extent in `ol.tilecoord.wrapX()` and `ol.tilecoord.clipX()`.

### PR DESCRIPTION
**Pull request by ahocevar**
This adds/fixes support for wrapping the date line for WMTS sources.

Fixes #3386.

### CODE REVIEW

File: externs/olx.js [hunk 1]
```diff
@@ -5690,6 +5691,14 @@ olx.source.WMTSOptions.prototype.urls;
 
 
 /**
+ * Whether to wrap the world horizontally. Default is `false`.
```
Comment (bartvde) [inline externs/olx.js]: what is your rationale for having false as the default?

File: externs/olx.js [hunk 2]
```diff
@@ -6502,7 +6512,9 @@ olx.tilegrid.TileGridOptions.prototype.origins;
 
 
 /**
- * Resolutions.
+ * Resolutions. The array index of each resolution needs to match the zoom
+ * level. This means that even if a `minZoom` is configured, the resolutions
+ * array will have a length of `maxZoom + 1`.
```
Comment (bartvde) [inline externs/olx.js]: is this something that should be documented for the release notes?

**Comment by bartvde** [comment]
I've put in two questions @ahocevar but other than that this LGTM thanks for your continued efforts on this

File: externs/olx.js (hunk 1 above)
Comment (ahocevar) [inline externs/olx.js]: WMTS almost always uses a custom tile grid. If that is not a global one and users don't specify widths for each resolution, things will go wrong. When generating a WMTS layer from capabilities, the default does not matter anyway because wrapX will be correctly set.

File: externs/olx.js (hunk 2 above)
Comment (ahocevar) [inline externs/olx.js]: It's not a change, just a clarification in the docs.

File: externs/olx.js (hunk 1 above)
Comment (bartvde) [inline externs/olx.js]: Thanks for the explanation, makes sense.

File: externs/olx.js (hunk 2 above)
Comment (bartvde) [inline externs/olx.js]: Ah okay. Then it's all good IMHO.

**Comment by ahocevar** [comment]
Thanks for the review.

File: examples/wmts-layer-from-capabilities.js [hunk 3]
```diff
@@ -24,7 +21,6 @@ $.ajax('data/WMTSCapabilities.xml').then(function(response) {
       }),
       new ol.layer.Tile({
         opacity: 1,
-        extent: projectionExtent,
```
Comment (elemoine) [inline examples/wmts-layer-from-capabilities.js]: Why did you remove this? Just wondering…

File: externs/olx.js (hunk 1 above)
Comment (elemoine) [inline externs/olx.js]: I agree with `wrapX` defaulting to `false` for WMTS.

File: examples/wmts-layer-from-capabilities.js (hunk 3 above)
Comment (ahocevar) [inline examples/wmts-layer-from-capabilities.js]: Setting an extent on the layer always prevents it from being rendered in multiple worlds.

File: externs/olx.js [hunk 4]
```diff
@@ -6527,12 +6539,22 @@ olx.tilegrid.TileGridOptions.prototype.tileSizes;
 
 
 /**
+ * Widths in tiles. If given, the array length should match the length of the
+ * `resolutions` array, i.e. each resolution will have a number of tile columns.
```
Comment (elemoine) [inline externs/olx.js]: Wouldn't it be good to explain when specifying this is necessary? (It's not clear to me for the moment.)

File: examples/wmts-layer-from-capabilities.js (hunk 3 above)
Comment (elemoine) [inline examples/wmts-layer-from-capabilities.js]: But in this example wrapX is not set to true.

**Comment by elemoine** [comment]
Thanks for the patch @ahocevar. I just allowed myself asking a few post-merge questions…

File: examples/wmts-layer-from-capabilities.js (hunk 3 above)
Comment (ahocevar) [inline examples/wmts-layer-from-capabilities.js]: ol.source.WMTS.optionsFromCapabilities sets `wrapX` to `true`.

File: externs/olx.js (hunk 4 above)
Comment (ahocevar) [inline externs/olx.js]: I'm working on a pull request to add an explanation. Thanks for asking.

File: examples/wmts-layer-from-capabilities.js (hunk 3 above)
Comment (elemoine) [inline examples/wmts-layer-from-capabilities.js]: Oh, I had missed that. Thanks!

File: externs/olx.js (hunk 4 above)
Comment (ahocevar) [inline externs/olx.js]: @elemoine, see #3399.
